File: modules/platforms/dotnet/Apache.Ignite.Core/Impl/Client/Cluster/ClientClusterGroupProjection.cs
/*
 * Copyright 2019 GridGain Systems, Inc. and Contributors.
 * 
 * Licensed under the GridGain Community Edition License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache.Ignite.Core.Impl.Client.Cluster
{
    using System;
    using System.Collections.Generic;
    using Apache.Ignite.Core.Binary;

    /// <summary>
    /// Projection builder that is used for remote nodes filtering.
    /// </summary>
    internal sealed class ClientClusterGroupProjection
    {
        /** */
        private const int Attribute = 1;

        /** */
        private const int ServerNode = 2;

        /** Filter value mappings. */
        private readonly Dictionary<int, object> _filter;

        private ClientClusterGroupProjection(Dictionary<int, object> filter)
        {
            _filter = filter;
        }

        /// <summary>
        /// Creates a new projection instance with specified attribute.
        /// </summary>
        /// <param name="name">Attribute name.</param>
        /// <param name="value">Attribute value.</param>
        /// <returns>Projection instance.</returns>
        public ClientClusterGroupProjection ForAttribute(string name, string value)
        {
            var filter = new Dictionary<int, object>(_filter);
            object attributes;
            if (filter.TryGetValue(Attribute, out attributes))
            {
                ((Dictionary<string, string>) attributes)[name] = value;
            }
            else
            {
                filter[Attribute] = new Dictionary<string, string> {{name, value}};
            }

            return new ClientClusterGroupProjection(filter);
        }

        /// <summary>
        /// Creates a new projection with server nodes only.
        /// </summary>
        /// <returns>Projection instance.</returns>
        public ClientClusterGroupProjection ForServerNodes()
        {
            var filter = new Dictionary<int, object>(_filter);
            filter[ServerNode] = true;
            return new ClientClusterGroupProjection(filter);
        }

        /// <summary>
        /// Initializes an empty projection instance.
        /// </summary>
        public static ClientClusterGroupProjection Empty
        {
            get { return new ClientClusterGroupProjection(new Dictionary<int, object>()); }
        }

        /// <summary>
        /// Writes the projection to output buffer.
        /// </summary>
        /// <param name="writer">Binary writer.</param>
        public void Write(IBinaryRawWriter writer)
        {
            if (_filter.Count == 0)
            {
                writer.WriteBoolean(false);
                return;
            }

            writer.WriteBoolean(true);
            writer.WriteInt(_filter.Count);

            foreach (var item in _filter)
            {
                switch (item.Key)
                {
                    case Attribute:
                    {
                        writer.WriteShort(Attribute);
                        var attributes = (Dictionary<string, string>) item.Value;
                        writer.WriteInt(attributes.Count);
                        foreach (var attr in attributes)
                        {
                            writer.WriteString(attr.Key);
                            writer.WriteString(attr.Value);
                        }

                        break;
                    }
                    case ServerNode:
                    {
                        writer.WriteShort(ServerNode);
                        break;
                    }
                    default:
                        throw new NotSupportedException(string.Format("Unknown filter code: {0}", item.Key));
                }
            }
        }
    }
}

File: modules/core/src/main/java/org/apache/ignite/internal/processors/platform/client/cluster/ClientClusterGroupProjection.java
/*
 * Copyright 2019 GridGain Systems, Inc. and Contributors.
 *
 * Licensed under the GridGain Community Edition License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.internal.processors.platform.client.cluster;

import org.apache.ignite.binary.BinaryRawReader;
import org.apache.ignite.cluster.ClusterGroup;

import java.util.function.Function;

/**
 * Client cluster group projection representation.
 * Decodes a remote projection request from a client node.
 */
public class ClientClusterGroupProjection {
    /** */
    private static final short ATTRIBUTE = 1;

    /** */
    private static final short SERVER_NODES = 2;

    /** Cluster group factory method. */
    private final Function<ClusterGroup, ClusterGroup> clusterGrpFactory;

    /**
     * Constructor.
     *
     * @param clusterGrpFactory Cluster Group builder.
     */
    private ClientClusterGroupProjection(Function<ClusterGroup, ClusterGroup> clusterGrpFactory){
        this.clusterGrpFactory = clusterGrpFactory;
    }

    /**
     * Reads projection from a stream.
     * @param reader Reader.
     * @return Projection.
     */
    public static ClientClusterGroupProjection read(BinaryRawReader reader) {
        if (!reader.readBoolean())
            return new ClientClusterGroupProjection(null);

        Function<ClusterGroup, ClusterGroup> factory = clusterGrp ->
        {
            int cnt = reader.readInt();
            for (int i = 0; i < cnt; i++) {
                short code = reader.readShort();
                switch (code) {
                    case ATTRIBUTE: {
                        String attName = reader.readString();
                        String attrVal = reader.readString();
                        clusterGrp = clusterGrp.forAttribute(attName, attrVal);
                        break;
                    }
                    case SERVER_NODES: {
                        clusterGrp = reader.readBoolean()
                                ? clusterGrp.forServers()
                                : clusterGrp.forClients();
                        break;
                    }
                    default:
                        throw new UnsupportedOperationException("Unknown code: " + code);
                }
            }
            return clusterGrp;
        };
        return new ClientClusterGroupProjection(factory);
    }

    /**
     * Applies projection.
     *
     * @param clusterGrp Source cluster group.
     * @return New cluster group instance with the projection.
     */
    public ClusterGroup apply(ClusterGroup clusterGrp){
        if(clusterGrpFactory == null)
            return clusterGrp;

        return clusterGrpFactory.apply(clusterGrp);
    }
}
